cmd/finch/gen_docs.go

```diff
@@ -0,0 +1,231 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package main
+
+import (
+	"bytes"
+	_ "embed"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+	"text/template"
+
+	"github.com/spf13/afero"
+	"github.com/spf13/cobra"
+
+	"github.com/runfinch/finch/pkg/flog"
+	"github.com/runfinch/finch/pkg/system"
+)
+
+// GenDocsSystemDeps contains the system dependencies for newGenDocsCommand.
+//
+//go:generate mockgen -copyright_file=../../copyright_header -destination=../../pkg/mocks/gen_docs_system_deps.go -package=mocks -mock_names GenDocsSystemDeps=GenDocsSystemDeps -source=gen_docs.go GenDocsSystemDeps
+type GenDocsSystemDeps interface {
+	system.PipeGetter
+	system.StdoutGetter
+	system.StdoutSetter
+}
+
+func newGenDocsCommand(
+	rootCmd *cobra.Command,
+	logger flog.Logger,
+	fs afero.Fs,
+	deps GenDocsSystemDeps,
+) *cobra.Command {
+	genDocsCommand := &cobra.Command{
+		Use:   "gen-docs",
+		Short: "Document generation",
+	}
+	genDocsCommand.AddCommand(
+		newGenDocsGenerateCommand(rootCmd, logger, fs, deps),
+	)
+	return genDocsCommand
+}
+
+func newGenDocsGenerateCommand(
+	rootCmd *cobra.Command,
+	logger flog.Logger,
+	fs afero.Fs,
+	deps GenDocsSystemDeps,
+) *cobra.Command {
+	genDocsGenerateCommand := &cobra.Command{
+		Use:    "generate",
+		Args:   cobra.NoArgs,
+		Hidden: true,
+		Short:  "Generate Finch docs",
+		RunE:   newGenDocsGenerateAction(rootCmd, logger, fs, deps).runAdapter,
+	}
+
+	genDocsGenerateCommand.Flags().StringP("path", "p", "", "Doc output directory")
+	// Ignore error since we check if the flag is set anyway
+	_ = genDocsGenerateCommand.MarkFlagRequired("path")
+
+	return genDocsGenerateCommand
+}
+
+type genDocsAction struct {
+	rootCmd *cobra.Command
+	logger  flog.Logger
+	fs      afero.Fs
+	deps    GenDocsSystemDeps
+}
+
+func newGenDocsGenerateAction(
+	rootCmd *cobra.Command,
+	logger flog.Logger,
+	fs afero.Fs,
+	deps GenDocsSystemDeps,
+) *genDocsAction {
+	return &genDocsAction{
+		rootCmd: rootCmd,
+		logger:  logger,
+		deps:    deps,
+		fs:      fs,
+	}
+}
+
+func (gd *genDocsAction) runAdapter(cmd *cobra.Command, _ []string) error {
+	path, err := cmd.Flags().GetString("path")
+	if err != nil {
+		return fmt.Errorf("failed to get required parameter 'path': %w", err)
+	}
+	return gd.run(path)
+}
+
+func (gd *genDocsAction) run(outDir string) error {
+	return gd.captureHelpOutput(gd.rootCmd, outDir)
+}
+
+func (gd *genDocsAction) captureHelpOutput(cmd *cobra.Command, outDir string) error {
+	for _, c := range cmd.Commands() {
+		if err := gd.captureHelpOutput(c, outDir); err != nil {
+			return err
```

Comment:
nit: adding command that resulted in the err would help with debug.
```suggestion
			return fmt.Errorf("error while generating docs for %s: %w", c.CommandPath(), e)
```

Reply:
> nit: adding command that resulted in the err would help with debug.

The reason I didn't wrap that error is because all errors that occur inside the `captureHelpOutput` function are already wrapped, so it would lead to double wrapping. Do you think it's still necessary?

Reply:
Decided that wrapping the error doesn't really hurt anything, updated in latest commit